File: scrimage-core/src/main/scala/com/sksamuel/scrimage/canvas/drawable/Polygon.scala
package com.sksamuel.scrimage.canvas.drawable

import com.sksamuel.scrimage.canvas.{Drawable, GraphicsContext}
import java.awt.Graphics2D
import scala.language.implicitConversions

case class Polygon(points: Seq[java.awt.Point], context: GraphicsContext) extends Drawable {
  def draw(g: Graphics2D): Unit = g.drawPolygon(points.map(_.x).toArray, points.map(_.y).toArray, points.size)
  def fill: FilledPolygon = FilledPolygon(points, context)
  def open: Polyline = Polyline(points, context)
}

object Polygon {

  implicit def awt2polygon(awt: java.awt.Polygon, context: GraphicsContext): Polygon = {
    val points = awt.xpoints.zip(awt.ypoints).map { case (x, y) => new java.awt.Point(x, y) }.toIndexedSeq
    Polygon(points, context)
  }

  def rectangle(x: Int, y: Int, width: Int, height: Int, context: GraphicsContext): Polygon = {
    Polygon(
      Seq(
        new java.awt.Point(x, y),
        new java.awt.Point(x + width, y),
        new java.awt.Point(x + width, y + height),
        new java.awt.Point(x, y + height)
      ),
      context
    )
  }
}

File: scrimage-core/src/main/java/com/sksamuel/scrimage/canvas/drawables/Polygon.java
package com.sksamuel.scrimage.canvas.drawables;

import com.sksamuel.scrimage.canvas.Drawable;
import com.sksamuel.scrimage.canvas.GraphicsContext;
import com.sksamuel.scrimage.canvas.RichGraphics2D;

import java.util.Arrays;
import java.util.List;

public class Polygon implements Drawable {

   public static Polygon rectangle(int x, int y, int width, int height, GraphicsContext context) {
      return new Polygon(
         Arrays.asList(
            new java.awt.Point(x, y),
            new java.awt.Point(x + width, y),
            new java.awt.Point(x + width, y + height),
            new java.awt.Point(x, y + height)
         ),
         context
      );
   }

   private final List<java.awt.Point> points;
   private final GraphicsContext context;

   public Polygon(List<java.awt.Point> points, GraphicsContext context) {
      this.points = points;
      this.context = context;
   }

   @Override
   public void draw(RichGraphics2D g) {
      int[] xs = points.stream().mapToInt(p -> p.x).toArray();
      int[] ys = points.stream().mapToInt(p -> p.y).toArray();
      g.drawPolygon(xs, ys, points.size());
   }

   public FilledPolygon toFilled() {
      return new FilledPolygon(points, context);
   }

   public Polyline toLine() {
      return new Polyline(points, context);
   }

   @Override
   public GraphicsContext context() {
      return context;
   }
}
